### PR TITLE
#1015 Select/deselect multiple rows in table using shift key

### LINKED ISSUE #1015: Select multiple rows in table using shift key
- Select one row, hold shift and select the last row - and all rows in between are automatically selected. This is especially necessary when all rows have to be included/ discarded at once. 
- Without selecting any row, if I shift + select nth row, all rows from 1-n will be selected

### PR DESCRIPTION
**Pull request by nmgokhale**
Fixes #1015 

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx [hunk 1]
```diff
@@ -88,7 +89,12 @@ class VirtualizedTable extends React.Component {
 					"index": index,
 					"originalRowIndex": rowData.originalRowIndex,
 					"selected": selected,
-					"isOverSelectOption": this.isOverSelectOption }, evt);
+					"isOverSelectOption": this.isOverSelectOption,
+					"selectMultipleRows": evt.shiftKey ? evt.shiftKey : false,
+					"lastCheckedRow": this.lastChecked === null ? 0 : this.lastChecked }, evt);
+
+				// Track lastChecked row for shift key selection
+				this.lastChecked = rowData.index; // originalRowIndex or index??
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:97]: Use originalRowIndex or index??

File: canvas_modules/common-canvas/src/common-properties/components/flexible-table/flexible-table.jsx [hunk 2]
```diff
@@ -329,10 +354,13 @@ class FlexibleTable extends React.Component {
 		if (!this.props.data[displayedRowIndex].disabled) {
 			if (overSelectOption) { // Checkbox is clicked
 				let current = this.props.selectedRows ? this.props.selectedRows : [];
-				if (checked) {
+				if (data.selectMultipleRows) { // multiple rows selected/deselected using shift key
+					current = this.handleCheckedMultipleRows(data.lastCheckedRow, displayedRowIndex, checked); // displayedRowIndex or dataRowIndex???
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/common-properties/components/flexible-table/flexible-table.jsx:358]: Use displayedRowIndex or dataRowIndex???

Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/components/flexible-table/flexible-table.jsx:358]: What is the difference?

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx (hunk 1 above)
Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:97]: What is the difference?

**Comment by matthoward366** [comment]
shift-click

I'm seeing an odd behavior where nothing is selected but then if I use shift-click it uses the last selection as the starting point

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx (hunk 1 above)
Comment (nmgokhale) [reply canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:97]: I didn't find any difference between `originalRowIndex` and `index`. I will keep the current code of using `index` for now.

File: canvas_modules/common-canvas/src/common-properties/components/flexible-table/flexible-table.jsx (hunk 2 above)
Comment (nmgokhale) [reply canvas_modules/common-canvas/src/common-properties/components/flexible-table/flexible-table.jsx:358]: `displayedRowIndex` is the index. And `dataRowIndex` is the originalRowIndex. Same comment as above. I will keep the current code of using `displayedRowIndex`  for now.

**Comment by nmgokhale** [comment]
> I'm seeing an odd behavior where nothing is selected but then if I use shift-click it uses the last selection as the starting point

This is expected. After opening a table for 1st time, without selecting any row when I shift+select a row --> All rows from 1-selected row will be checked.
But after opening a table if any row is manually selected/deselected, shift + select starts from last row selection.

**Pull request merged**